File: my-folder/0042-trapping-rain-water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:

        maxleft, maxright = 0, 0
        left, right = 0, len(height)-1
        trapwater = 0
        while left < right:  

            if height[left] <= height[right]:
                maxleft = max(maxleft, height[left])
                trapwater += maxleft - height[left]
                left += 1
            else:
                maxright = max(maxright, height[right])
                trapwater += maxright - height[right]
                right -= 1
        
        return trapwater
        """
        ml = 0, mr = 1

        """

            

File: my-folder/0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = height.length-1;
        int left_max = 0;
        int right_max = 0;
        int trapped_water = 0;
        while (left < right){
            if (height[left] < height[right]){
                left_max = Math.max(left_max, height[left]);
                trapped_water += left_max - height[left++];
            } else{
                right_max = Math.max(right_max, height[right]);
                trapped_water += right_max - height[right--];
            }
        }
        return trapped_water;

    }
}
